File: graph/dijkstra/dijkstra.py
from heapq import *
import sys
input = sys.stdin.readline
INF = int(1e6)

def dijkstra(start):
    heap = []
    heappush(heap, (0, start))
    distance[start] = 0
    while heap:
        dist, now = heappop(heap)
        if distance[now] < dist:
            continue
        for i in graph[now]:
            cost = dist + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heappush(heap, (cost, i[0]))

n, m = map(int, input().split())
graph = [[] for _ in range(n+1)]
distance = [INF] * (n+1)

for _ in range(m):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))
    graph[b].append((a, c)) # 양방향일 때 추가

start = int(input())
dijkstra(start)

File: graph/dijkstra/dijkstra.cpp
#include <bits/stdc++.h>
#define fastio cin.tie(0), cout.tie(0), ios_base::sync_with_stdio(false)
#define MAX 20001
#define INF 1234567890
typedef long long ll;
using namespace std;

ll v, e, a, b, c, k, d, now, cost;
vector<vector<pair<ll, ll>>> graph(MAX);
ll dist[MAX];
priority_queue<pair<ll, ll>> pq;

void dijkstra(ll start){
    pq.push({0, start});
    dist[start] = 0;
    while (!pq.empty()){
        d = -pq.top().first; now = pq.top().second;
        pq.pop();
        if (dist[now] < d) continue;
        for (auto nxt: graph[now]){
            cost = d+nxt.second;
            if (cost < dist[nxt.first]){
                dist[nxt.first] = cost;
                pq.push({-cost, nxt.first});
            }
        }
    }
}

int main(){
    fastio;
    cin >> v >> e >> k;
    fill(dist, dist+MAX, INF);
    for (ll i=0; i<e; i++){
        cin >> a >> b >> c;
        graph[a].push_back(make_pair(b, c));
    }
    dijkstra(k);
    for (ll i = 1; i <= v; i++){
        if (dist[i] != INF) cout << dist[i];
        else cout << "INF";
        cout << '\n';
    }
}
